File: leetcode/easy/222_完全二叉树的节点个数/222_完全二叉树的节点个数.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

//递归
func getNodesNum(cur *TreeNode) int {
	if cur == nil {
		return 0
	}
	leftNum := getNodesNum(cur.Left)
	rightNum := getNodesNum(cur.Right)
	treeNum := leftNum + rightNum + 1
	return treeNum
}

func countNodes(root *TreeNode) int {
	return getNodesNum(root)
}



//迭代法
func countNodes(root *TreeNode) int {
	que := []*TreeNode{}
	if root != nil {
		que = append(que, root)
	}

	result := 0
	for len(que) > 0 {
		size := len(que)
		for i := 0; i < size; i++ {
			node := que[0]
			que = que[1:]
			result++
			if node.Left != nil {
				que = append(que, node.Left)
			}
			if node.Right != nil {
				que = append(que, node.Right)
			}
		}
	}
	return result
}

File: leetcode/easy/222_完全二叉树的节点个数/222_完全二叉树的节点个数.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

//递归
class Solution {
private:
    int getNodesNum(TreeNode* cur) {
        if(cur==0) return 0;
        int leftNum=getNodesNum(cur->left);
        int rightNum=getNodesNum(cur->right);
        int treeNum=leftNum+rightNum+1;
        return treeNum;
    }
public:
    int countNodes(TreeNode* root) {
        return getNodesNum(root);
    }
};



//迭代法
class Solution {
public:
    int countNodes(TreeNode* root) {
        queue<TreeNode*>que;
        if(root!=NULL) que.push(root);
        int result=0;
        while(!que.empty()) {
            int size=que.size();
            for(int i=0;i<size;i++)
            {
                TreeNode* node = que.front();
                que.pop();
                result++;
                if(node->left) que.push(node->left);
                if(node->right) que.push(node->right);
            }
        }
        return result;
    }
};